Main.hs

```diff
@@ -106,38 +106,29 @@ convertMove m s r
    | m `elem` "rR" = Discard (convertCard s r)
    | otherwise     = error "Unrecognized move"
 
-readCards :: IO [Card]
-readCards = readCards' []
-   where
-   readCards' :: [Card] -> IO [Card]
-   readCards' cs = do
-      s <- getChar
-      if s == '.'
-         then do getChar -- to get new line character
-                 return cs
-         else do r <- getChar
-                 getChar -- to get new line character
-                 let c = convertCard s r
-                 readCards' (cs ++ [c])
+readCards :: IO CardList
+readCards = do line <- getLine
+               if line == "."
+                  then return []
+                  else do rest <- readCards
+                          return ([validateAndConvertCard line] ++ rest) where
+                            validateAndConvertCard :: String -> Card
+                            validateAndConvertCard [c1,c2] = convertCard c1 c2
+                            validateAndConvertCard _       = error "invalid string for card"
 
 readMoves :: IO [Move]
-readMoves = readMoves' []
-   where
-   readMoves' :: [Move] -> IO [Move]
-   readMoves' ms = do
-      m <- getChar
-      if m == '.'
-         then do getChar -- to get new line character
-                 return ms
-         else if m `elem` "dD"
-              then do let m' = convertMove m ' ' ' '
-                      getChar  -- to get new line character
-                      readMoves' (ms ++ [m'])
-              else do s <- getChar
-                      r <- getChar
-                      getChar -- to get new line character
-                      let m' = convertMove m s r
-                      readMoves' (ms ++ [m'])
+readMoves = do line <- getLine
+               if line == "."
+                  then return []
+                  else do rest <- readMoves
+                          return ([validateAndConvertMove line] ++ rest) where
+                            err :: a
+                            err = error "invalid string for move"
+
+                            validateAndConvertMove :: String -> Move
```

Comment:
Again splitting this function into two would simplify this function dramatically.